// foxie-framework/src/gulpfile.js
var gulp = require('gulp'),
		gulpFoxie = require('gulp-foxie'),
		gulpStylus = require('gulp-stylus'),
		stylus = require('stylus'),
		livereload = require('gulp-livereload'),
		browserify = require('gulp-browserify'),
		map = require('./map.js'),
		uglify = require('gulp-uglify'),
		dirSync = require('gulp-dir-sync'),
		plugins = require('gulp-load-plugins')()

function foxieFramework(){
	this.styles = function(){
		return gulpFoxie.styles(gulp, map, gulpStylus, livereload);
	}
	this.scripts = function(){
		return gulpFoxie.scripts(gulp, map, browserify, uglify, livereload);
	}
	this.json = function(){
		return gulpFoxie.json(gulp, map, livereload);
	}
	this.watch = function(){
		var targets = {
			styles:map.styles.recursive.styl,
			scripts:map.scripts.recursive.js,
			settings:map.styles.recursive.json
		}
		return targets;
	}

	this.syncPaths = function(){
		var targets = {
			settingsSrc:map.sync.src.settings,
			settingsDest:map.sync.dest.settings
		}
		return targets;
	}
	this.sync = function(){
		dirSync(map.sync.src.settings, map.sync.dest.settings);
	}

	this.default = function(){
		var tasksList = Object.keys(gulp.tasks);
		return tasksList;
	}
	this.entry = map.styles.main;
}
foxieFramework = new foxieFramework();

//Standalone dev server
gulp.task("foxie-styles", foxieFramework.styles());
gulp.task("foxie-scripts", foxieFramework.scripts());
//gulp.task("foxie-sync", foxieFramework.json());

gulp.task("foxie-watch", function(){
	//[build server function] watch is unrelated to the exportable targets function seen above.
	gulp.watch(map.styles.recursive.styl,['foxie-styles']);
	gulp.watch(map.scripts.recursive.js,['foxie-scripts']);
	//gulp.watch(map.styles.recursive.json, ['foxie-sync']);
});
//run all tasks that have been defined
gulp.task("default", foxieFramework.default());


module.exports = foxieFramework;
